server/src/main/java/org/elasticsearch/index/query/RangeQueryBuilder.java

```diff
@@ -438,6 +438,7 @@ public String getWriteableName() {
     protected MappedFieldType.Relation getRelation(final CoordinatorRewriteContext coordinatorRewriteContext) {
         final MappedFieldType fieldType = coordinatorRewriteContext.getFieldType(fieldName);
         if (fieldType instanceof final DateFieldMapper.DateFieldType dateFieldType) {
+            assert fieldName.equals(fieldType.name());
             IndexLongFieldRange fieldRange = coordinatorRewriteContext.getFieldRange(fieldName);
             if (fieldRange.isComplete() == false || fieldRange == IndexLongFieldRange.EMPTY) {
```

Comment:
The diff does not include it, but the core of the issue is that we return DISJOINT for an EMPTY range, hence we skip. I wonder if we should rather change this: if we have incomplete data should we rather let the query run instead of skipping and taking the risk of not returning matching data like in this case? On the other hand if EMPTY reliably means empty shard, it makes sense to skip. Not too sure about the isComplete part here.


Also, I cannot think of other situations where timestamp comes back as the right field type and checks all the boxes for coord rewrite, yet perhaps we were not able to store the range in the cluster state, because min and max were not available. In that case we'd end up skipping. It may be safer to not skip instead. It would have saved us from this bug at least.

Reply:
I think this the relevant discussion about that issue from a PR I had that was closed not merged: 
https://github.com/elastic/elasticsearch/pull/110873#pullrequestreview-2177808205
https://github.com/elastic/elasticsearch/pull/110873#issuecomment-2228671729
https://github.com/elastic/elasticsearch/pull/110873#issuecomment-2259189944

Reply:
And Armin had the same question here: https://github.com/elastic/elasticsearch/pull/111419#discussion_r1695669970
But I never answered him on the PR. I think I referred him to the above discussion outside the PR. I didn't want to change functionality more than necessary since the changes around event.ingested were so complex and I was worried of subtle bugs.

But now might be a time to raise a GH issue around this and get another discussion going.

Reply:
> the core of the issue is that we return DISJOINT for an EMPTY range, hence we skip. I wonder if we should rather change this: if we have incomplete data should we rather let the query run instead of skipping

`EMPTY` doesn't (or at least shouldn't) mean "incomplete data" - its docs say:

> Sentinel value indicating an empty range, for instance because the field is missing or has no values in any shard.

Thus `DISJOINT` should be right. If however there's some bug that meant we incorrectly record some indices as having `EMPTY` range when in fact we meant `UNKNOWN` then I would recommend coming up with some fix for that rather than changing the meaning of `EMPTY` forever.

Reply:
that's indeed the purpose of this PR. And that's what we discussed a few pixels above. 

I will keep this bit the same as it is. The reason why I wonder is that it is much much worse to not return matching data than doing more roundtrips in very specific situations, like empty shards. Anyways, I will leave that for another time.